sockeye/decoder.py

```diff
@@ -286,23 +282,24 @@ def forward(self, step_input, states):
     def hybrid_forward(self, F, step_input, states):
         mask = None
         if self.inference_only:
-            steps, source_mask, *other = states
+            steps, att_valid_length, *other = states
             source_encoded = None  # use constant pre-computed key value projections from the states
             enc_att_kv = other[:self.config.num_layers]
             autoregr_states = other[self.config.num_layers:]
         else:
             if any(layer.needs_mask for layer in self.layers):
                 mask = self.autoregressive_bias(step_input)  # mask: (1, length, length)
-            steps, source_encoded, source_mask, *autoregr_states = states
+            steps, source_encoded, att_valid_length, *autoregr_states = states
             enc_att_kv = [None for _ in range(self.config.num_layers)]
 
         if any(layer.num_state_tensors > 1 for layer in self.layers):
             # separates autoregressive states by layer
             states_iter = iter(autoregr_states)
             autoregr_states = [list(islice(states_iter, 0, layer.num_state_tensors)) for layer in self.layers]
 
-        # Fold the heads of source_mask (batch_size, num_heads, seq_len) -> (batch_size * num_heads, 1, seq_len)
-        source_mask = F.expand_dims(F.reshape(source_mask, shape=(-3, -2)), axis=1)
+        # (batch_size * heads, query_length)
+        att_valid_length = layers.prepare_softmax_lengths(F, att_valid_length, step_input,
```

Comment:
I can guess what this function does, but its name is a bit ambiguous.

Reply:
Do you have a better suggestion? I can't think of a better alternative.

Reply:
I can only think of a boring name `prepare_source_valid_lengths` :)

Reply:
ok will do